backend/pkg/serde/service.go

```diff
@@ -0,0 +1,282 @@
+	return &Service{
+		SerDes: []Serde{
+			NoneSerde{},
+			JSONSerde{},
+			JSONSchemaSerde{SchemaSvc: schemaService},
+			XMLSerde{},
+			AvroSerde{SchemaSvc: schemaService},
+			ProtobufSerde{ProtoSvc: protoSvc},
+			ProtobufSchemaSerde{ProtoSvc: protoSvc},
+			MsgPackSerde{MsgPackService: msgPackSvc},
+			SmileSerde{},
+			UTF8Serde{},
+			TextSerde{},
+			UintSerde{},
+		},
+	}
+}
+
+// DeserializeRecord tries to deserialize a Kafka record into a struct that
+// can be processed by the Frontend.
+func (s *Service) DeserializeRecord(ctx context.Context, record *kgo.Record, opts DeserializationOptions) *Record {
+	// defaults
+	if opts.MaxPayloadSize <= 0 {
+		opts.MaxPayloadSize = defaultMaxPayloadSize
+	}
+
+	// 1. Test if it's a known binary Format
+	if record.Topic == "__consumer_offsets" {
+		rec, err := s.deserializeConsumerOffset(record)
+		if err == nil {
+			return rec
+		}
+	}
+
+	// 2. Deserialize key & value separately
+	key := s.deserializePayload(ctx, record, PayloadTypeKey, &opts)
+	val := s.deserializePayload(ctx, record, PayloadTypeValue, &opts)
+	headers := recordHeaders(record)
+
+	return &Record{
+		Key:     key,
+		Value:   val,
+		Headers: headers,
+	}
+}
+
+// deserializePayload deserializes either the key or value of a Kafka record by trying
+// the pre-defined deserialization strategies.
+func (s *Service) deserializePayload(ctx context.Context, record *kgo.Record, payloadType PayloadType, opts *DeserializationOptions) *RecordPayload {
+	payload := payloadFromRecord(record, payloadType)
+
+	var originalPayload []byte
+	if opts.IncludeRawData {
+		originalPayload = payload
+	}
+	// Check if payload is empty
+	if len(payload) == 0 {
+		return &RecordPayload{
+			OriginalPayload:  originalPayload,
+			IsPayloadNull:    payload == nil,
+			PayloadSizeBytes: 0,
+			Encoding:         PayloadEncodingNone,
+		}
+	}
+
+	troubleshooting := make([]TroubleshootingReport, 0)
+
+	serdeEncoding := opts.KeyEncoding
+	if payloadType == PayloadTypeValue {
+		serdeEncoding = opts.ValueEncoding
+	}
+
+	doCustomEncoding := serdeEncoding != PayloadEncodingUnspecified && serdeEncoding != ""
+
+	// Try all registered SerDes in the order they were registered
+	var rp *RecordPayload
+	var err error
+	for _, serde := range s.SerDes {
+		if doCustomEncoding {
+			if serdeEncoding != serde.Name() {
+				continue
+			}
+		}
+
+		rp, err = serde.DeserializePayload(ctx, record, payloadType)
+		if err == nil {
+			// found the matching serde
+			break
+		}
+
+		troubleshooting = append(troubleshooting, TroubleshootingReport{
+			SerdeName: string(serde.Name()),
+			Message:   err.Error(),
+		})
+	}
+
+	addTS := opts.Troubleshoot
+	if rp == nil || err != nil || (rp != nil && rp.Encoding == "") {
+		// Anything else is considered binary
+		rp = &RecordPayload{
+			Encoding: PayloadEncodingBinary,
+		}
+
+		// if we failed to find a successful serde always add troubleshooting info
+		addTS = true
+	}
```

Comment:
Would it be cleaner to add a default serde at the very end of the serde list? Not sure, but the if condition feels a bit hard to understand / maintain.